Alias

nriker opened the pull request:
Реализованы builtins Alias и Unalias, а также api для хеш-таблиц (последнее - хедер t_hashtable.h)
Более подробное описание с ссылками на источники во вкладке Issues.

ИЛЬЯ: тебе чтоб получать alias нужно использовать функцию char	*alias_api_get_alias(char *key); из хедера t_hashalias.h

Прогонял валгриндом на маке и убунту

nriker committed:
Начало разработки alias
Создана архитектура API hash-таблиц и хеш-таблиц для alias

nriker committed:
Реализована хеш-функция.
Процесс реализации api для hashtable.

nriker committed:
Реализовано API для создания таблицы, добавления элементов, изменения элементов в таблице
Реализация API для удаления элементов и t_hashdate

nriker committed:
Создано API для хе-таблицы
Разработка api для alias/unalias

nriker committed:
Пофикшен баг вывода автоподстановки: перемешивался старый вывод и промт

nriker committed:
Нормировка хедеров

nriker committed:
Реализована печать алиасов

nriker committed:
Удалены лишние файлы
Сделать добавление alias
Сделать удаление alias

nriker committed:
Добавлен в main статическое объявление alias
Добавлено получение alias по ключу

nriker committed:
Реализована вставка alias

nriker committed:
Добавлен built in unalias
Прогнать в valgrind

nriker committed:
Убраны из вариантов подстановок . и ..

nriker committed:
Valgrind. Mistakes

nriker committed:
Исправлен alias

ag95v2 reviewed (changes requested):
1) Залей, пожалуйста, это в ветку 42_unstable.

2) в этот PR попали изменения в несколько файлов из readline/completion. Ясное дело, к alias они отношения не имеют. Разбей, плз, на 2 PR, если это сильно разные фичи

3) hashtable.h - api выглядит немного перегруженным.
Если у нас есть такая функция static_hashalias_action то может просто оставить только такой api?
char *search(char *key), 
void insert(char *key, char *value), 
void delete(char *key) + print_table, если нужно?
Они под капотом будут юзать static_hashalias_action,  пользователь ничего не узнает. Очень хочется спрятать все ненужное от глаз читателя.

Имя hash_api_delete_hash_date выглядит очень странным, date - дата.

4) есть вопросы к hash_api_print_hash_table. Зачем она выводит приглашение и вообще почему зависит от readline?

5) hash_api_init_hash_data делает memalloc, а потом memset нулями. memalloc уже делает memset. И ты в обеих ветках if возвращаешь 1 и то же))

6) Мы с Артуром в 21sh решили, что непрошедший malloc обрабатывать в этом проекте нецелесообразно, можно юзать xmalloc, который будет завершать программу. Это может сильно сократить объем кода.

7) builtin alias предполагает в твоем случае, что в качестве аргументов передается несколько строк типа а = б
По-моему, там должна быть одна строка со знаком = посередине типа a=b, причем = должен быть не заэкранирован и вообще имя должно удовлетворять определенным условиям. Этот момент хорошо бы уточнить в стандарте.

volkov7 commented:
Я так понял, этот код будет совместим только с веткой dev, было бы неплохо сделать функционал совместимый с веткой 42_unstable, потому что там актуальная версия шелла

volkov7 commented on src/hashtable/hash_api_hash_function.c:
```diff
@@ -6,14 +6,27 @@
+		key++;
+	}
+	hash = hash % HASH_SIZE;
+	return ((unsigned int)hash);
```
Функция возвращает unsigned int, а в сигнатуре функции int. Еще не совсем нравится, что для вычисления используется unsigned long, а затем происходит сужающее преобразование, мне кажется, это нехорошо

volkov7 commented on src/hashtable/hash_api_insert_hash_data.c:
```diff
@@ -0,0 +1,64 @@
+	if ((new->key = ft_strdup(key)) == NULL
+		|| (new->value = ft_strdup(value)) == NULL)
+	{
+		free(new);
```
Тут не совсем логически верно, если для ключа память выделится, а для value нет, то мы очистим указатель new, а указатель для key потеряем -> утечка. Думаю правильней будет написать функцию, которая будет принимать элемент хэш-таблицы и очищать ненулевые поля, а затем сам выделенный элемент и возвращать статус EXIT_FAILURE

nriker committed:
Изменены названия функций для alias api и hashtable api.
Изменено возвращаемое значение hash_function на unsigned_int
Добавлены проверки при free в insert_hash_date
Удалена бесполезная печать gaypromt в print_hash_table
